### PR TITLE
Clean up Molecule API docs

### PR DESCRIPTION
**Pull request by Yoshanuikabundi**
I've gone over the Book-style docs and the Molecule API docs to fill in any missing or outdated info and correct any style mistakes.

I've also added a bunch of type annotations. These are especially useful for documenting properties! I've hit every property in the Molecule class except those that have units - the typing story for quantity-wrapped numpy arrays isn't quite there yet.

- [x] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/master/docs), if applicable
- [x] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [ ] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/master/docs/releasehistory.rst)

### CODE REVIEW

**Review by mattwthompson — APPROVED** [review]
LGTM, I made a few non-blocking drive-by comments but overall I am thrilled to see so many of these changes coming in. I'd be happy to give more feedback on any details later if you'd like. But now I'm still hitting the approve button (pending tests passing).

I'll admit I didn't look through `docs/users/molecule_conversion.md` in detail, nor did I look at the rendered docs, since that would take a few hours to do thoroughly. But the changes I saw certainly **look** like they'd produce cleanly rendered docs.

I saw a lot of wrapping to 80 characters, some changes being just wrapping to that. I'm pretty used to wrapping code to 100-120 and I don't personally stick to the classic 80x25 terminal size. Is there a particular reason for these changes, or is it something that was automatically done by a tool? All else being equal I have a personal preference for wider docstrings that roughly match the width of the wall of code, or just keeping them however they were when they were written. (Well, I guess what I'd _love_ to see is a pre-commit hook that automatically wraps everything to some number of lines around 100 so that everything is always committed consistently from the beginning. But I have not looked into that at all and it certainly doesn't need to be done here!)

File: openff/toolkit/topology/molecule.py
Comment (Yoshanuikabundi) [reply openff/toolkit/topology/molecule.py]: Yeah RST is a nightmare, the obvious thing never works. I'm totally spoiled by MyST

**Comment by Yoshanuikabundi** [comment]
> I saw a lot of wrapping to 80 characters, some changes being just wrapping to that. I'm pretty used to wrapping code to 100-120 and I don't personally stick to the classic 80x25 terminal size. Is there a particular reason for these changes, or is it something that was automatically done by a tool?

It's just [PEP8](https://peps.python.org/pep-0008/#maximum-line-length). Black targets a width of 79 characters but allows lines to go to 99 characters if necessary, following PEP8. PEP8 says "For flowing long blocks of text with fewer structural restrictions (docstrings or comments), the line length should be limited to 72 characters" which is... extreme. So 79 characters is a compromise believe it or not!

I'm happy to do whatever here. 70-80 characters is about optimal for readability and is what I target for paragraph widths in the rendered docs. I tend to find 100 characters a bit long for prose and 80 is about right once you have an indent, and it being PEP8 standard-ish is a bonus. But it would be great to standardise this across OpenFF as I like 80 characters, you like 100, and Simon likes 120, so moving across projects involves a lot of config changes for me (or more accurately, a lot of me imposing my will on everyone else's docs).

File: openff/toolkit/topology/molecule.py
Comment (Yoshanuikabundi) [reply openff/toolkit/topology/molecule.py:4638]: It's actually to get the function to type check - from inspecting the source `_construct_torsions` always sets `_propers` to a set (though it may be empty). The assert just tells mypy that. I prefer asserts to just silencing the type checker because if something changes you get a crash that should show up in tests. I like asserts as a way to document invariants, but I'd consider any assert that actually raises an exception a bug - it means that an invariant that you're relying on is not being upheld. I'd rather an exception in that case than an incorrect result. I agree that they're not a good way to actually check if something's true, and if a user sees one something's gone wrong.

I'm happy to add a message though :)

File: openff/toolkit/topology/molecule.py
Comment (Yoshanuikabundi) [reply openff/toolkit/topology/molecule.py]: Thanks, fixed

File: openff/toolkit/topology/molecule.py
Comment (Yoshanuikabundi) [reply openff/toolkit/topology/molecule.py:5492]: I don't know. It's not in 0.10.6 so hopefully it gets removed before 0.11.0 is released 😕

**Comment by mattwthompson** [comment]
Of those options I would prefer 120 or 100 width. In terms of source code, have `flake8` configured to yell at me if I go past that (inherited from the [cookitcutter's settings](https://github.com/MolSSI/cookiecutter-cms/blob/161869500ea39fa60342d400e74beea4fa2ac6dc/%7B%7Bcookiecutter.repo_name%7D%7D/setup.cfg), which I'm mostly happy with). I could probably wrap to 100 if there was a drive to do so, but I think the cat's out of the 80 character bag at this point. And I only bring up source code line lengths because I think it looks a bit out of place to have docstrings much narrower than the code they're inserted inside of. Docstrings at 100 work okay, but long-ish variable names (i.e. [what I have up now](https://github.com/openforcefield/openff-interchange/blob/b3ab49cc60ee019147b978ae6bd965d88a9e0071/openff/interchange/components/interchange.py#L214) or the various `ToolkitRegistry`/`ToolkitWrapper` lines you wrestled with here) don't work well with 80 characters to me.

All this being said - I'd be willing to compromise in order to advance consistency. And this doesn't need to be acted on here!

**Review by j-wags — APPROVED** [review]
You don't need my approval since @mattwthompson already gave it, but I really like a lot of what you did here - Especially on the molecule conversion page!

**Comment by j-wags** [comment]
Yeah, the line width issue doesn't need action here. For the record my vote is for 80. It's not based in any good reason, 80 is just... the number of columns.

**Pull request merged**
